### ISSUE TITLE
Feature request: custom fallback action for narrowed source

### ISSUE THREAD

**Issue by chasecaleb** [minad/consult#539]
TLDR: Is there a way to define a fallback action for when `consult-buffer` is narrowed to a specific source and no candidates match the minibuffer input? Or would you be willing to add support for this somehow?

---

In https://github.com/minad/consult#multiple-sources you have an example of defining a dummy source to create a new buffer, like this:

```elisp
(defun mode-buffer-exists-p (mode)
  (seq-some (lambda (buf)
              (provided-mode-derived-p
               (buffer-local-value 'major-mode buf)
               mode))
            (buffer-list)))

(defvar eshell-source
  `(:category 'consult-new
    :face     'font-lock-constant-face
    :action   ,(lambda (_) (eshell))
    :items
    ,(lambda ()
       (unless (mode-buffer-exists-p 'eshell-mode)
         '("*eshell* (new)")))))
```

Doing that is nice, but it would be even better if there was a way to make that "*eshell* (new)"  candidate show up regardless of what is typed into the minibuffer.

Here's the workflow that this would enable:
1. Run `consult-buffer`
2. Narrow to a specific source (for instance, I have a source defined for EXWM Firefox buffers)
3. Start typing to find a buffer, but nothing matches because e.g. I forgot that I closed the buffer
4. Fallback candidate is still shown regardless of what is typed into minibuffer, allowing me to hit enter to create a new buffer using a custom action (like the `(eshell)` lambda in the snippet above).

Unfortunately as best as I can tell, there isn't a way of providing a fallback candidate/action like I refer to in step 4... instead, if I hit enter when there are no matches shown in `consult-buffer` then it will always create a new `fundamental-mode` buffer.

**Comment by minad** [comment]
These fallback candidates are a nice idea. The implementation is not difficult. See this demo, where the ALWAYS candidates are always present and the FALLBACK candidates are only present if no other candidates match.

~~~ elisp
(let ((coll '("first" "second" "third")))
  (completing-read "Prompt: "
                   (lambda (str pred action)
                     (pcase action
                       (`(boundaries . ,_) nil)
                       ('metadata nil)
                       ('nil (try-completion str coll pred))
                       ('t (append (or (all-completions str coll pred)
                                       (list (format "FALLBACK1: %s" (minibuffer-contents))
                                             (format "FALLBACK2: %s" (minibuffer-contents))))
                                   (list (format "ALWAYS1: %s" (minibuffer-contents))
                                         (format "ALWAYS2: %s" (minibuffer-contents)))))
                       (_ (test-completion str coll pred))))))
~~~

Before adding a feature like this, I'd like to think about a few more use cases where such a functionality could be useful. Are there existing Consult commands besides `consult-buffer` (with narrowing as you described) where fallback (or always) candidates could come handy?

**Comment by chasecaleb** [comment]
Nice demo, I like it. 

I waited a bit before responding to think about other uses, but haven't come up with anything compelling beyond the switch-to-or-create buffer scenario. But some extra examples that come to mind regarding buffer switching: EXWM, vterm, other shells, magit, dired, org agenda... just about any sort of interactive buffer that isn't backed by a file.

The one not-so-great idea I came up with was for `consult-find`, `consult-grep`, and other similar search type commands: you could add a fallback action to change the search directory if there are no matches. I don't think that's a very compelling use though, especially with how easy it is to just re-run the command with a `C-u` prefix then hit `M-p` to re-enter the same query.

---

I think it would make sense to keep it specific to `consult-buffer` and allow configuring the fallback as a property on buffer sources. For example, I have a consult source like this for Firefox:

``` elisp
(:name     "Web"
 :narrow   (?w . "Web")
 :hidden   t
 :category buffer
 :face     consult-buffer
 :history  buffer-name-history
 :state    ,#'consult--buffer-state
 :items
 ,(lambda ()
    (consult--buffer-query
     :sort 'visibility
     :predicate
     (lambda (buffer)
       (and (equal (buffer-local-value 'major-mode buffer) 'exwm-mode)
            (equal (buffer-local-value 'exwm-class-name buffer) "Firefox")))
     :as #'buffer-name)))
```
It would be convenient to add something along the lines of a `:fallback` property specifying the completion text and an action function to invoke when selected.

**Comment by minad** [comment]
> I waited a bit before responding to think about other uses, but haven't come up with anything compelling beyond the switch-to-or-create buffer scenario.

Okay, I don't think it is a priority to add this then. I like the idea but it is better to wait for more input. Maybe we come up with more use cases. The point is that the Consult API is layered, consult-buffer builds on consult--multi, which builds on consult--read. This change would affect all of these layers. For now I don't want to make an addition which is only useful for the topmost layer.

**Comment by minad** [comment]
Closing. We can revisit this if we come up with more compelling use cases. While I find the idea neat, I am also critical about this since these FALLBACK or ALWAYS candidates deviate from how completion usually works and it is an edge case feature anyway. Therefore the barrier for adding the additional complexity of such a feature is a bit higher.

**Comment by minad** [comment]
I plan to implement this soon such that sources can create new candidates if they have a `:new` field. The active source will be selected by narrowing as proposed by @chasecaleb.

**Comment by minad** [comment]
Sources can have a `:new` function now which is called for non-existing candidates. This can only happen if the `consult--multi` option `:require-match` is nil. 

However I haven't glued narrowing to the `:new` functionality. If `consult-buffer` is narrowed for example to an Eshell buffer source, then entering a new buffer name should create a new Eshell buffer (as defined by the `:new` function of the source).

**Comment by minad** [comment]
I finished the implementation in https://github.com/minad/consult/commit/8bccc73073247cd395a84165515a68468790a6dd. Unfortunately the change comes with a minor API change of `consult--read`. This affects these two MELPA packages:

* [consult-flyspell](https://gitlab.com/OlMon/consult-flyspell/-/blob/master/consult-flyspell.el#L104) by @OlMon
* [consult-org-roam](https://github.com/jgru/consult-org-roam/blob/da86614508e5d138b517e25b0752611563e3117e/consult-org-roam.el#L150) by @jgru

Is there a reason why you don't use one of the predefined `consult--lookup-*` functions? I hope we can keep the `consult--read` API more stable from now on.

**Comment by oantolin** [comment]
For consult-flyspell @OlMon should probably just use cons pairs (i.e. add a `.` to line 85 😄) and the standard `consult--lookup-cdr).

**Comment by minad** [comment]
@oantolin Thanks. I think in most cases the predefined lookup functions should suffice. The only truly special ones we have in Consult are `consult--line-match` and `consult--multi-lookup`.

**Comment by minad** [comment]
I made the change. I also reordered the arguments by importance: selected, candidates, input, narrow. Most functions only need the first two. Only the aforementioned special lookup functions need narrow and input.

**Comment by oantolin** [comment]
> @oantolin Thanks. I think in most cases the predefined lookup functions should suffice. The only truly special ones we have in Consult are `consult--line-match` and `consult--multi-lookup`.

consult-org-roam has a `:lookup` function that returns the input verbatim if not found in the candidate alist, so it looks like @jgru does need a custom lookup function.

**Comment by jgru** [comment]
> I finished the implementation in [8bccc73](https://github.com/minad/consult/commit/8bccc73073247cd395a84165515a68468790a6dd). Unfortunately the change comes with a minor API change of `consult--read` regarding the `:lookup` function. This affects these two MELPA packages:
>     * [consult-org-roam](https://github.com/jgru/consult-org-roam/blob/da86614508e5d138b517e25b0752611563e3117e/consult-org-roam.el#L150) by @jgru
> 
> 
> Is there a reason why you don't use one of the predefined `consult--lookup-*` functions? I hope we can keep the `consult--read` API more stable from now on.

Thanks for the heads-up, @minad.

> consult-org-roam has a `:lookup` function that returns the input verbatim if not found in the candidate alist, so it looks like @jgru does need a custom lookup function.

Indeed, I implemented the custom function to take and handle the input as it is. This is required to capture the current input on `vertico-exit-input`. BTW, happy that you had a look at my code, @oantolin. 

Given this need, I decided to adapt my custom function according to the API-change in [e7787a2](https://github.com/jgru/consult-org-roam/commit/e7787a272835a2df7185b333bc4bb08e542bfd01).  I hope that this is fine. 

Best regards
jgru

**Comment by minad** [comment]
@chasecaleb See the `org-source` example in https://github.com/minad/consult#multiple-sources, where I make use of the `:new` function.

**Comment by OlMon** [comment]
Hej Hej,

@minad and @oantolin thanks for notifying and giving a solution! 
I changed consult-flyspell.

Cheers, 
keep up the good work.

**Comment by chasecaleb** [comment]
@minad _Finally_ got a chance to use this and wanted to say thank you! It solves my needs perfectly.
